src/libcore/iter.rs

```diff
@@ -1690,6 +1702,22 @@ impl<B, I: Iterator, F> Iterator for Map<I, F> where F: FnMut(I::Item) -> B {
     fn size_hint(&self) -> (usize, Option<usize>) {
         self.iter.size_hint()
     }
+
+    #[inline]
+    fn count(self) -> usize {
+        self.iter.count()
+    }
+
+    #[inline]
+    fn last(self) -> Option<B> {
+        let f = self.f;
+        self.iter.last().map(f)
```

Comment:
Because it doesn’t compile this way. `self` is by value here and is therefore partially moved by `self.iter` which does not allow capturing `self` into the closure.